### PR TITLE
Adaptive jbuf

### PR DESCRIPTION
**Pull request by cspiel1**
audio, stream: put to and get from jitter buffer in two threads
    
    During a call there is the RTP receive thread (A) and the auplay thread (B).
    Previously the two threads had been connected by an aubuf. Thread A put the
    RTP packets in the jbuf and also did a get in the same thread. Thus jbuf did
    not really work as jitter buffer. Thread A did also the decoding and pushed the
    data into the aubuf. Thread B fetched the data from the aubuf and pushed it to
    the audio device.
    
aubuf-2threads

    Now in libre the jbuf was made thread safe and it works now as adaptive jitter
    buffer. What means that the latency is controlled in reaction on the measured
    network jitter.

Thread B has real time constraints and decoding could be to time consuming (e.g. on slow HW). We added a third thread X for decoding. Like shown in this image: 
jbuf-3threads
    
    
    Adapting of the latency in the jbuf is done during periods of silence. This
    commit also adds a silence detection. Which is disabled by default because it
    is the only O(n) complexity part, where n is the number of samples (sampc).

Related to: https://github.com/baresip/re/pull/41.

### CODE REVIEW

**Comment by premultiply** [comment]
Just a note:
Received (encoded) RTP packets should be never dropped but decoded audio samples may be dropped or silence may be added if necessary.

The aubuffer is always read with the clock of the local audio output device. So it must underrun or overflow from time to time if the audio output clock does not exactly match the senders input audio clock.
The jitterbuffer does make shure that also delayed packets, packet bursts and reordered packets can be decoded in line.

**Comment by sreimers** [comment]
> So the three thread version would not have any benefits.

In my use case the auplay thread uses samples from 64 to 960 @48khz, i need to test if opus decoding is fast enough to
decode packets with ptime 20ms with deadline 64 samples (real time window lower than 1,3ms@48kHz).

**Comment by alfredh** [comment]
please note that the so called "jitter buffer" has a misleading name, and should
in reality be called a "reordering" buffer to reorder incoming UDP/RTP packets.

the auplay handler code must be very fast, and only copy audio samples from a buffer.

there is a chance that auplay handler may run in the context of a real-time thread,
see Audiounit or Portaudio documentation.

File: src/audio.c [hunk 1]
```diff
@@ -1105,9 +1127,9 @@ int audio_alloc(struct audio **ap, struct list *streaml,
 
 	err = stream_alloc(&a->strm, streaml,
 			   stream_prm, &cfg->avt, sdp_sess,
-			   MEDIA_AUDIO, label,
+			   MEDIA_AUDIO, label, true, ptime,
```
Comment (alfredh) [inline src/audio.c:1130]: I dont think that `ptime` should be sent into stream.
ptime is audio specific, and stream is generic.

File: src/core.h [hunk 2]
```diff
@@ -289,6 +289,7 @@ enum {STREAM_PRESZ = 4+12}; /* same as RTP_HEADER_SIZE */
 typedef void (stream_rtp_h)(const struct rtp_header *hdr,
 			    struct rtpext *extv, size_t extc,
 			    struct mbuf *mb, unsigned lostc, void *arg);
+typedef int (stream_pt_h)(uint8_t pt, struct mbuf *mb, void *arg);
```
Comment (alfredh) [inline src/core.h:292]: is it possible to make a separate PR for this change, including the code that is
handling payload-type changes?

File: src/stream.c [hunk 3]
```diff
@@ -502,10 +524,10 @@ int stream_alloc(struct stream **sp, struct list *streaml,
 		goto out;
 
 	/* Jitter buffer */
-	if (cfg->jbuf_del.min && cfg->jbuf_del.max) {
+	if (jbuf && cfg->jbuf_del.min && cfg->jbuf_del.max) {
 
 		err = jbuf_alloc(&s->jbuf, cfg->jbuf_del.min,
-				 cfg->jbuf_del.max);
+				 cfg->jbuf_del.max, cfg->jbuf_wish, ptime);
```
Comment (alfredh) [inline src/stream.c:530]: is there a need to send `ptime` to jbuf ?

File: src/video.c [hunk 4]
```diff
@@ -897,9 +893,10 @@ int video_alloc(struct video **vp, struct list *streaml,
 	tmr_init(&v->tmr);
 
 	err = stream_alloc(&v->strm, streaml, stream_prm,
-			   &cfg->avt, sdp_sess, MEDIA_VIDEO, label,
+			   &cfg->avt, sdp_sess, MEDIA_VIDEO, label, false, 0,
```
Comment (alfredh) [inline src/video.c:896]: the video stream also needs a jitter/reordering buffer ?

File: src/stream.c (hunk 3 above)
Comment (cspiel1) [reply src/stream.c:530]: Yes, the adaptive jitter buffer algorithm needs ptime. See libre PR! Earlier I had a version where ptime was somehow computed from the RTP timestamps. But it was not very reliable.

File: src/video.c (hunk 4 above)
Comment (cspiel1) [reply src/video.c:896]: Maybe yes. So I should drop the bool jbuf from stream_alloc()?

File: src/audio.c (hunk 1 above)
Comment (cspiel1) [reply src/audio.c:1130]: The ptime is passed to jbuf. It is the time difference between two RTP packets. For audio it is the same like the frame size. For Video it is the time how long a video frame is viewed.
In this sense, the packet time is stream specific. No matter if video or audio. Right?

**Comment by cspiel1** [comment]
> > So the three thread version would not have any benefits.
> 
> In my use case the auplay thread uses samples from 64 to 960 @48kHz, i need to test if opus decoding is fast enough to
> decode packets with ptime 20ms with deadline 64 samples (real time window lower than 1,3ms@48kHz).

You mean that jbuf_put() is called much more often then jbuf_get()?
I think, you have seen already: The stream_decode() is called on demand and in a while loop until the num_bytes is reached in the auplay write handler.

Our assumption is: If the decoding (e.g. opus) is not fast enough it would also be to slow for the current jbuf and same for the three thread version. At least for single core CPUs. But we can discuss about a three threaded version.

File: src/core.h (hunk 2 above)
Comment (cspiel1) [reply src/core.h:292]: I'll try so. Anyway, I forgot to call s->pth() during preparing the PR.

Comment (cspiel1) [reply src/core.h:292]: Currently try to divide the big commit into:
- pt-handler
- silence detection
- stream decode

File: src/audio.c [hunk 5]
```diff
@@ -629,6 +631,27 @@ static void check_telev(struct audio *a, struct autx *tx)
...
+	}
+#else
+	(void) sampv;
+	(void) sampc;
+#endif
+
+	return true;
+}
+
+
```
Comment (alfredh) [inline src/audio.c:654]: would it be possible to use this API for silence detection:

`double aulevel_calc_dbov(int fmt, const void *sampv, size_t sampc);`

it supports more audio sample formats, and it also calculates the
correct energy/power level of the signal.

**Comment by alfredh** [comment]
after some thinking I am wondering if we should use the new jbuf only for the audio stream.
the video stream can still use the current jbuf.

also the new code can be placed in baresip/src/. and given a new name.

it might also be possible to add a config item to select the jbuf type used.

The ideal receive pipeline for incoming audio:

```
[ RTP ] -> [ Reordering buffer ] -> [ Decode ] -> [ Adaptive Playout buffer ] -> [ auplay ]
```

The new logic that you have implemented, should be done on the decoded audio samples.

The re-ordering buffer can be e.g. 2-4 slots, and all incoming RTP packets must be decoded.
It can also be skipped if you know that the network has perfect UDP transport.

As @premultiply said, all rtp packets must be decoded.

**Comment by premultiply** [comment]
The "reordering buffer" is more or less a timeout waiting for a delayed, reordered or lost packet which should be as long as the maximum expected jitter to avoid skipping late packets but handle real lost packets on the receiver side by continuing with the next received rtp packet in the buffer in proper order.

**Comment by cspiel1** [comment]
Now pushed again:
- Activated jbuf also for video stream. Tested already.
- ptime is computed from RTP timestamps and works reliable now. So stream_alloc() does'nt need the ptime parameter.
- Split the commit into parts: pt-handler, jbuf with two threads, wish size, silence detection

**Comment by cspiel1** [comment]
> after some thinking I am wondering if we should use the new jbuf only for the audio stream.
> the video stream can still use the current jbuf.
> 
> also the new code can be placed in baresip/src/. and given a new name.
> 
> it might also be possible to add a config item to select the jbuf type used.
> 
> The ideal receive pipeline for incoming audio:
> 
> ```
> [ RTP ] -> [ Reordering buffer ] -> [ Decode ] -> [ Adaptive Playout buffer ] -> [ auplay ]
> ```
> 
> The new logic that you have implemented, should be done on the decoded audio samples.
> 
> The re-ordering buffer can be e.g. 2-4 slots, and all incoming RTP packets must be decoded.
> It can also be skipped if you know that the network has perfect UDP transport.
> 
> As @premultiply said, all rtp packets must be decoded.

- Another buffer means more latency. My solution has the benefit that reorder and jitter control is done in one buffer. The minimum value is to guarantee that reordering can be done. And the maximum value can be chosen relatively high and let the jbuf extend if a high jitter is determined. If the jbuf increases the number of stored packets when the network jitter increases, also the reordering can be done more effectively. 
- Dropping RTP packets is only done to reduce latency if the network condition improves. It is also and already done for late out-of-order packets, for old packets in case of an overflow and anyway it is also possible that packets are lost during the transmission. Anyway, the decoder works also with missing packets.
- The adaptive jitter buffer algorithm needs the RTP header (timestamp, seq-nr), thus the decoded frames would be to less information for computing the network jitter. Please see the libre PR!

A customer of our VOIP products reported audio artifacts and we found that a high network jitter was the problem. But increasing the jitterbuffer minimum value was not applicable because of the high latency. In coordination with our audio specialists I developed the first version of the adaptive jbuf in June 2019. I also compared with gstreamer. Gstrtpjitterbuffer also works with RTP packets as input and output. Also gstreamer has to drop too old packets. After a few improvements we have a stable version in productive systems since this year February.

**Comment by cspiel1** [comment]
Added a commit that avoids packet drops in jbuf.

File: src/stream.c [hunk 6]
```diff
@@ -505,7 +542,7 @@ int stream_alloc(struct stream **sp, struct list *streaml,
 	if (cfg->jbuf_del.min && cfg->jbuf_del.max) {
 
 		err = jbuf_alloc(&s->jbuf, cfg->jbuf_del.min,
-				 cfg->jbuf_del.max);
+				 cfg->jbuf_del.max, cfg->jbuf_wish);
```
Comment (alfredh) [inline src/stream.c:545]: perhaps avoid breaking the api by adding a new function, e.g.:

```c
void jbuf_set_wish(struct jbuf *jb, uint32_t wish);
```

Comment (cspiel1) [reply src/stream.c:545]: I'll add a commit to avoid the push force. Maybe later we should make a interactive rebase before merge.

**Comment by alfredh** [comment]
I would like to suggest that you make a separate PR for the stream pt handler code.

this PR is quite big, and it is difficult to review with many unrelated changes.

as a first step, please make a PR only for the pt handler.
then we merge it to master, and rebase this patch.

File: src/audio.c [hunk 7]
```diff
@@ -643,9 +668,22 @@ static void check_telev(struct audio *a, struct autx *tx)
...
+	while (err == EAGAIN ||
+			(!err && aubuf_cur_size(rx->aubuf) < num_bytes)) {
+
+		if (err == EAGAIN)
+			++rx->again;
+
+		err = stream_decode(a->strm);
+	}
+
```
Comment (alfredh) [inline src/audio.c:686]: we should not do audio decode in the audio play handler.

this handler may be called from audio DSP realtime interrupt-handler.

I suggest instead to do decoding in:

1. a new thread
2. a `struct tmr` polling loop

**Comment by alfredh** [comment]
the PT handler patch was merged to master.

please rebase this PR on top of master :)

File: src/audio.c (hunk 7 above)
Comment (cspiel1) [reply src/audio.c:686]: Ok, I also agree and will add a commit that does one of this suggestions.

Comment (cspiel1) [reply src/audio.c:686]: I think both with #ifdef HAVE_PTHREAD, right?

Found that the union around the "pthread_t tid" is obsolete because the AUDIO_MODE_TMR was removed. But I could use the original union approach to select tmr vs thread. Ok?

**Comment by cspiel1** [comment]
Tested with different network jitter values. Tested also the tmr version and the thread version.
jbuf-jitter-50ms

**Comment by cspiel1** [comment]
Some of the commits should be squashed before merging.

File: src/audio.c [hunk 8]
```diff
@@ -172,8 +172,20 @@ struct aurx {
...
+		struct {
+			pthread_t tid;
+			bool run;
+			pthread_cond_t cond;
+			pthread_mutex_t mutex;
+		} thr;                /**< Audio decode thread             */
+#else
+		struct tmr tmr;       /**< Timer for audio decoding        */
+#endif
+	} u;
```
Comment (sreimers) [inline src/audio.c:188]: I think with ifdef/else there is only one union member at time, so the union looks a bit needless?

Comment (cspiel1) [reply src/audio.c:188]: Yes. I'll remove this. Same for the struct autx? I ask already 3 days ago.

File: src/audio.c [hunk 9]
```diff
@@ -629,6 +661,88 @@ static void check_telev(struct audio *a, struct autx *tx)
...
+	struct aurx *rx = &a->rx;
+	int err = 0;
+
+	while (rx->wcnt > 0) {
+		err = stream_decode(a->strm);
+		rx->wcnt--;
+	}
+
+	while (err == EAGAIN ||
+			(!err && aubuf_cur_size(rx->aubuf) < rx->num_bytes)) {
```
Comment (sreimers) [inline src/audio.c:698]: I have segfault with call hangup:
```
#0  0x00007fc9c2d02a76 in pthread_rwlock_rdlock () at /usr/lib/libpthread.so.0
#1  0x000055bff209f1b0 in lock_read_get (l=0xb5b5b5b5b5b5b5b5) at src/lock/rwlock.c:66
#2  0x000055bff20add0a in aubuf_cur_size (ab=0x55bff2c78d80) at src/aubuf/aubuf.c:339
#3  0x000055bff1fe5c35 in audio_decode (arg=0x55bff2c695f0) at src/audio.c:698
#4  0x000055bff1fe5d70 in rx_thread (arg=0x55bff2c695f0) at src/audio.c:738
#5  0x00007fc9c2cfe3e9 in start_thread () at /usr/lib/libpthread.so.0
#6  0x00007fc9bf162293 in clone () at /usr/lib/libc.so.6
```

Comment (cspiel1) [reply src/audio.c:694]: Tried hangup on both sides. For the moment I am not able to reproduce this. It seems that the decoding thread was not stopped cleanly. I should replace two lines in stop_rx(). Like this:
```
		rx->thr.run = false;
		pthread_cond_signal(&rx->thr.cond);
```

Comment (sreimers) [reply src/audio.c:694]: Looks good now. Thanks.

Comment (cspiel1) [reply src/audio.c:694]: Ok, also saw the segfault if it was an outgoing call.

**Comment by sreimers** [comment]
with rx_thread I see a very high CPU usage (100% instead of 10% without patches), I haven't looked further (or test tmr), let me know if you need debug information.

**Comment by cspiel1** [comment]
Thanks! Again this is only for outgoing call. I am able to reproduce this.

**Comment by cspiel1** [comment]
Had to move the
```
	pthread_mutex_destroy(&a->rx.thr.mutex);
	pthread_cond_destroy(&a->rx.thr.cond);
```
from stop_rx() to the destructor. Which makes e.g. a difference for outgoing calls.

**Comment by alfredh** [comment]
hi again Christian,

many thanks for your efforts, this is a very nice feature.

I apologise for my forth and back answers here, during the past couple of days I did some
more thinking about this. Please bear with me :)

my biggest worry is that this is a large change, and will break something that we did not think
about, or break something for some production users.

it is also a bit annoying that some code lives in baresip and some in re.

I therefore propose the following:

1. We move the jbuf module from re to baresip/src, perhaps with a new name (e.g. ajbuf)
2. We apply your patch on this new module
3. Add a config item to control which jbuf type is used:
```
enum jbuf_type {
    JBUF_OFF,
    JBUF_FIXED,
    JBUF_ADAPTIVE
};
```
this way users can choose to use the old jbuf method (fixed).

4. After merge, the development continues in baresip/src/ajbuf.c and we can delete re/src/jbuf

This way you can make atomic PR's in baresip that does not depend on other changes.

Please let me know what you think, it is just an idea from my side.

PS: Christian, do you also want to have commit-rights to the baresip project ?

**Comment by cspiel1** [comment]
Hi Alfread, indeed it is a change which goes deep into the core of baresip. I also think there could be done more testing with different codecs, different min/max/wish settings for jbuf and aubuf. Also different network behavior.

Making an ajbuf beside of jbuf and choosing between them is a good idea. But I am not sure and I have to check if it is possible to choose between old and new behavior of stream/audio.
Old behavior: The RTP thread does jbuf_put and jbuf_get and decoding. So also pushes into aubuf.
New behavior: The three threads method as shown in above figure.

On the other side for stream/video I think we still have the old behavior. So maybe a few if-conditions will be enough for stream/audio. So I have to check/try this and add a commit.

But more concretely to the ajbuf:
- The adaptive jbuf is useful for incoming RTP streams in general. Are you sure that we should move it from re to baresip?
- To your point 4.: You suggest to delete re/src/jbuf. Do you mean that from then on the old jbuf and the new ajbuf are living in baresip side-by-side?
- Did you also think about making a baresip module ajbuf? We could simply decide that it has to go into the baresip core. But we could also invent a new filter type and add a "register_jbuf" or "select_jbuf" or "set_jbuf" method for a jitter buffer. This would make it possible for module developer to add different jitter buffer implementations and then choose one of them during runtime.


Thanks that you want to grant me commit rights! The current workflow with PRs and reviews surely is advantageous for the quality. We have to re-think about the usefulness of our patches, improve our changes and try to deliver simple and understandable PRs. You and the other maintainers are very fast and responsible. Your suggestions are helpful and useful. What would be the benefit if I have commit rights and when should I make use of it?

regards, Christian.

File: src/audio.c [hunk 10]
```diff
@@ -112,12 +114,10 @@ struct autx {
...
-	union {
-		struct {
-			pthread_t tid;/**< Audio transmit thread           */
-			bool run;     /**< Audio transmit thread running   */
-		} thr;
-	} u;
+	struct {
+		pthread_t tid;/**< Audio transmit thread           */
+		bool run;     /**< Audio transmit thread running   */
+	} thr;
```
Comment (alfredh) [inline src/audio.c:120]: this change is unrelated, I fixed it in master:

4b8a6d4180252ba7662f66da37feecb6fb1bbc31

**Comment by cspiel1** [comment]
Hi Alfred, so the most important question we have to solve before I can proceed is if we should move the jitter buffer from re to baresip. Most annoying for me is that travis fails if the API of re changes. In this case the function jbuf_set_wish() was added. But @sreimers has a good plan to solve this in general for PRs. He wrote:

> It looks like there is already a nice github action for our needs: https://github.com/marketplace/actions/github-action-build-chain-cross-repo-builds (i will take a closer look at this)

So should we really move jbuf to baresip? Ok, it is above re_rtp.h. But usable for any incoming RTP stream. Not only for baresip. Maybe also @sreimers has an opinion about this? Who else?

**Comment by cspiel1** [comment]
Second question: Should jbuf be replace able by different implementations (like ausrc is)? I started to implement an adapter:
```
/** The jitter buffer adapter makes jitter buffer implementations exchange-
 * able.
 * */
struct jbuf_ad {
	struct jbuf *st;
	
	jbuf_set_wish_h * set_wish;
	jbuf_put_h *put;
	jbuf_get_h *get;
	jbuf_set_silence_h *set_silence;
	jbuf_flush_h *flush;
	jbuf_stats_h *stats;
	jbuf_debug_h *debug;
};
...
/* use like this: */
err = s->jbuf->put(s->jbuf->st, hdr, mb);
```
But then I stopped. It seems that to be too much flexibility, which we don't need. And with the old and the new jbuf living in parallel (in re or baresip that doesn't really matter for this question) it would be a lot of code duplication. 

Instead I suggest to add a new method to the new jbuf which select between the old and the new behavior.
```
jbuf_set_method(struct jbuf *jbuf, enum jbuf_type jbtype);
```

@alfredh, please let me know which approach you prefer!

Additionally as mentioned above stream/audio will get the configuration for the jbuf type.

**Comment by alfredh** [comment]
it should be as simple and straightforward as possible.

adding a jbuf abstraction I think is not needed, it is a bit overengineering.

a good solution is to add a jbuf type, that can be used to select the wanted behaviour:

```c
enum jbuf_type {
    JBUF_OFF,
    JBUF_FIXED,
    JBUF_ADAPTIVE
};
```

the default behaviour should be "FIXED".

**Comment by cspiel1** [comment]
Ok, I pushed in re/baresip the selectable jbuf type. Tested all three variants (with incoming/outgoing calls this time).

Let me know if I should move jbuf to baresip! The failing travis should not be the reason for that, because @sreimers is about to solve this.

**Comment by sreimers** [comment]
@cspiel1 I just merged my pull request dependency ci action code. Can you try to rebase from master? It should detect the baresip/re pull request (same github pr title) automatically.

https://github.com/baresip/baresip/pull/1132

**Comment by cspiel1** [comment]
@sreimers Yes! It works. Great solution!

**Comment by cspiel1** [comment]
@alfredh All tests are green now with the new github action of @sreimers . Selecting type JBUF_FIXED vs JBUF_ADAPTIVE work. Also in retest there are now two separate tests for fixed/adaptive. Do you still prefer that I move jbuf to baresip?

I let the PR to draft, because if you are ready with review I want to make a cleaner git history.

**Comment by alfredh** [comment]
thanks. I will take a look at this during the next couple of days ..

**Comment by cspiel1** [comment]
Is it ok, if I start the interactive rebase to squash some of the commits and then push force?

**Comment by alfredh** [comment]
yes, okay :)

**Comment by alfredh** [comment]
merged to master.

Many thanks to Christian and his team for this feature 👍

**Comment by cspiel1** [comment]
You're welcome! Thanks for your review!
